File: 27RemoveElement.go
package main

import (
	"fmt"
)

func removeElement(nums []int, val int) int {
	start := 0
	end := len(nums) - 1
	var output int

	for start <= end {
		if start == end {
			if nums[start] == val {
				output = start
			} else {
				output = start + 1
			}
			break
		}

		if nums[start] == val {
			tmp := nums[start]
			nums[start] = nums[end]
			nums[end] = tmp
			end--
		} else {
			start++
		}
	}

	return output
}

func main() {
	output := removeElement([]int{3, 2, 2, 3}, 3)
	output1 := removeElement([]int{0, 1, 2, 2, 3, 0, 4, 2}, 2)

	fmt.Println("\n******** output *********\n", output, output1)
}

File: 27RemoveElement.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
    return;

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        int len = nums.size();
        for (int i = 0; i < len;) {
            if (nums[i] == val) {
                swap(nums[i], nums[len-1]);
                len--;
            } else {
                i++;
            }
        }

        return len;
    }
};

int main(){
    Solution s;

    auto vecInput = vector<int>{3, 2, 2, 3};
    auto output = s.removeElement(vecInput, 3);

    auto vecInput1 = vector<int>{0, 1, 2, 2, 3, 0, 4, 2};
    auto output1 = s.removeElement(vecInput1, 2);

    cout << "*************output*************" << endl;
    cout << output << endl;
    cout << output1 << endl;

    return 0;
}
